File: C#/shakerSort.cs
using System;

class ShakerSort
{
    // Method to perform Shaker Sort on an array
    public static void Sort(int[] array)
    {
        int left = 0; // Starting index of the unsorted portion
        int right = array.Length - 1; // Ending index of the unsorted portion
        bool swapped; // Flag to check if a swap occurred

        do
        {
            swapped = false; // Reset the swap flag for this pass

            // Forward pass: Move the largest element to the right
            for (int i = left; i < right; i++)
            {
                // Compare adjacent elements and swap if they are in the wrong order
                if (array[i] > array[i + 1])
                {
                    Swap(array, i, i + 1); // Swap the elements
                    swapped = true; // Set flag to true since a swap occurred
                }
            }
            right--; // Decrease the right boundary since the last element is sorted

            // Backward pass: Move the smallest element to the left
            for (int i = right; i > left; i--)
            {
                // Compare adjacent elements and swap if they are in the wrong order
                if (array[i] < array[i - 1])
                {
                    Swap(array, i, i - 1); // Swap the elements
                    swapped = true; // Set flag to true since a swap occurred
                }
            }
            left++; // Increase the left boundary since the first element is sorted

        } while (swapped); // Continue until no swaps occur
    }

    // Method to swap two elements in the array
    private static void Swap(int[] array, int indexA, int indexB)
    {
        int temp = array[indexA]; // Store the value at indexA
        array[indexA] = array[indexB]; // Move the value at indexB to indexA
        array[indexB] = temp; // Assign the stored value to indexB
    }

    // Main method to test the Shaker Sort algorithm
    static void Main()
    {
        // Example array to be sorted
        int[] array = { 64, 34, 25, 12, 22, 11, 90 };

        Console.WriteLine("Original array:");
        PrintArray(array); // Print the original array

        // Call the sort method
        Sort(array); // Sort the array using Shaker Sort

        Console.WriteLine("Sorted array:");
        PrintArray(array); // Print the sorted array
    }

    // Method to print the elements of the array
    private static void PrintArray(int[] array)
    {
        foreach (int item in array)
        {
            Console.Write(item + " "); // Print each item followed by a space
        }
        Console.WriteLine(); // Print a new line after the array
    }
}

File: JavaScript/shakerSort.js
/**
 * Function to perform Shaker Sort on an array.
 * @param {number[]} arr - The array to be sorted.
 * @returns {number[]} - The sorted array.
 */
function shakerSort(arr) {
    let n = arr.length; // Get the length of the array
    let swapped; // Variable to track if a swap occurred
    let start = 0; // Starting index for the forward pass
    let end = n - 1; // Ending index for the backward pass

    do {
        swapped = false; // Reset swapped to false at the start of each pass

        // Forward pass: Move largest element to the end
        for (let i = start; i < end; i++) {
            if (arr[i] > arr[i + 1]) {
                // Swap if elements are in the wrong order
                [arr[i], arr[i + 1]] = [arr[i + 1], arr[i]];
                swapped = true; // Set swapped to true
            }
        }

        // If no swaps occurred, the array is sorted
        if (!swapped) break;

        // Decrease end since the last element is sorted
        end--;

        // Backward pass: Move smallest element to the beginning
        for (let i = end; i > start; i--) {
            if (arr[i] < arr[i - 1]) {
                // Swap if elements are in the wrong order
                [arr[i], arr[i - 1]] = [arr[i - 1], arr[i]];
                swapped = true; // Set swapped to true
            }
        }

        // Increase start since the first element is sorted
        start++;

    } while (swapped); // Continue if any swaps were made

    return arr; // Return the sorted array
}

// Example usage
const arrayToSort = [64, 34, 25, 12, 22, 11, 90];
const sortedArray = shakerSort(arrayToSort);
console.log("Sorted Array:", sortedArray);
